### DISCUSSION TITLE
union_all with literal potential bug

### DISCUSSION THREAD

**Discussion by ShmuelTreiger** [sqlalchemy/sqlalchemy#6714] [Usage Questions]
I believe this is a bug:

`session.execute(select(literal(True), Model).filter(...))`

Works fine, gives me (True, Object)

```
u1 = select(Model).filter(...)

u2 = select(Model).filter(...)

session.execute(select(union_all(u1, u2).subquery()))
```

Also works. However when I do:

```
u1 = select(literal(True), Model).filter(...)

u2 = select(literal(False), Model).filter(...)

session.execute(select(union_all(u1, u2).subquery()))
```

SQLA deconstructs the object from Model into its component pieces, so the result is:

`(bool <from the literal>, int, str, int, datetime, ...)`

I've looked through the documentation and couldn't find anything, but that's failed me before, so I thought I'd create a discussion before raising an issue. Appologies in advance if this is designed/documented behavior and I missed it.

**Comment by zzzeek** [comment]
selecting from a subquery or union() directly will always return individual columns, not ORM entities.  To get an ORM entity from subquery() or union() you need to select from an [aliased](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#orm-entity-subqueries-ctes).

your examples are invalid SQLAlchemy API calls and will not run as written as you appear to be embedding result sets in a union() (see error message below), so if you can produce an [MCVE](http://stackoverflow.com/help/mcve) that shows very exactly what you are doing such that I can run it here, I will have a better idea of what you are doing.


```py
from sqlalchemy import union_all

u1 = s.execute(select(A))

u2 = s.execute(select(A))

s.execute(select(union_all(u1, u2).subquery()))
```

output:

```
sqlalchemy.exc.ArgumentError: SELECT construct for inclusion in a UNION or other set construct expected, got <sqlalchemy.engine.result.ChunkedIteratorResult object at 0x7f363e924b80>```

**Reply by ShmuelTreiger** [reply]
Sorry. Edited. Just had to take out the `session.execute`'s

**Reply by zzzeek** [reply]
does not do what you describe.  Here's an MCVE.  Run this script:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import literal
from sqlalchemy import select
from sqlalchemy import union_all
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    x = Column(Integer)
    y = Column(Integer)
    z = Column(Integer)


e = create_engine("sqlite://", echo=False)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(x=2, y=3, z=4))
s.commit()


def test_one_i_return_a_model():
    return s.execute(select(literal(True), A).filter(A.id != None)).all()


def test_two_i_return_columns():
    u1 = select(A).filter(A.id != None)

    u2 = select(A).filter(A.id != None)

    return s.execute(select(union_all(u1, u2).subquery())).all()


def test_three_i_return_columns():
    u1 = select(literal(True), A).filter(A.id != None)

    u2 = select(literal(False), A).filter(A.id != None)

    return s.execute(select(union_all(u1, u2).subquery())).all()


for fn in (
    test_one_i_return_a_model,
    test_two_i_return_columns,
    test_three_i_return_columns,
):
    print(f"{fn.__name__}: {fn()}")
```

output:

```
$ python test3.py 
test_one_i_return_a_model: [(True, <__main__.A object at 0x7f3edbef41c0>)]
test_two_i_return_columns: [(1, 2, 3, 4), (1, 2, 3, 4)]
test_three_i_return_columns: [(True, 1, 2, 3, 4), (False, 1, 2, 3, 4)]
```

**Reply by ShmuelTreiger** [reply]
Sorry. I was working on it.

**Reply by ShmuelTreiger** [reply]
OK, I now understand that if I want the object after union, I need to do

`aliased(Obj, union)`

Is there a way to get (literal, Obj) after the union?

Sorry, but I can't figure it out.

**Reply by zzzeek** [reply]
you can use the union object multiple times in the columns clause

```py
def test_four_i_return_both():
    u1 = select(literal(True), A).filter(A.id != None)

    u2 = select(literal(False), A).filter(A.id != None)

    subq = union_all(u1, u2).subquery()

    return s.execute(select(subq.c[0], aliased(A, subq))).all()
```

**Reply by ShmuelTreiger** [reply]
This works and makes sense to me. Thank you.

Out of curiosity, why does it work like this? Specifically, when I instead do:

`return s.execute(select(aliased(A, subq))).all()`

It gives me rows containing the A objects. How does aliased know to select the right columns and ignore the others? Does it just search for the object it's trying to create from the given data columns?

Purely curiosity and trying to understand how SQLA works. Thank you.

**Reply by zzzeek** [reply]
wow , thanks for being the first person to ever ask me in 16 years!   the main way that the ORM figures out what columns are what is using a thing called "column correspondence", which I wrote a blog post about many years ago (hoping someone would tell me what it is I'm doing, or would have advice, or critique...nothing :)  )  https://techspot.zzzeek.org/2008/01/23/expression-transformations/  

in this script, we can see this as follows.  First we have the ``Table`` object that we've created a mapping towards:

```py
>>> A.__table__
Table('a', MetaData(), Column('id', Integer(), table=<a>, primary_key=True, nullable=False), Column('x', Integer(), table=<a>), Column('y', Integer(), table=<a>), Column('z', Integer(), table=<a>), schema=None)

# the "x" column
>>> A.__table__.c.x
Column('x', Integer(), table=<a>)
```

then we have the subquery/union thing we've made:

```py
>>> u1 = select(literal(True), A).filter(A.id != None)
>>> u2 = select(literal(False), A).filter(A.id != None)
>>> subq = union_all(u1, u2).subquery()
```

when we build out a construct like that, every time the column named "x" gets accessed to create a new SQL construct that can be SELECTed from, we make a copy of that column in terms of the new table; above it's generated from the subquery that we made based on the two unions:

```py
>>> subq.c.x
Column('x', Integer(), table=<anon_1>)
```

when this column was generated, it keeps track of where it came from:

```py
>>> subq.c.x._proxies
[Column('x', Integer(), table=<a>), Column('x', Integer(), table=<a>)]
```

(the column repeated twice has to do with the fact that the subquery is against a union that includes this column in two separate SELECT statements)

so by setting up this "proxies" relationship, we can track the "x" column between the `A.__table__` and the subquery in both directions; this is not based on the "name" of the column at all, it's based on how the subquery was generated from the table object and is all done structurally:

```py
# the subquery gives us the column that came from the "x" column on the table
>>> subq.c.corresponding_column(A.__table__.c.x)
Column('x', Integer(), table=<anon_1>)

# the table can give us which of its columns relate to this column in the subquery
>>> A.__table__.c.corresponding_column(subq.c.x)
Column('x', Integer(), table=<a>)
```

using that general mechanical feature, the ORM builds out a huge bunch of logic that I've been fixing bugs in for over a decade that knows how to relate the "selectable" to which `A` is mapped, i.e. the `Table`, to the columns that happen to be in this subquery , so that it knows which columns to put in the SELECT statement it renders and it knows how to get them back from a result set and populate them into a new `A` object.  and that's how the sausage is made.

**Reply by CaselIT** [reply]
You never stop learning I guess.

May be worth coping this reply in a faq item somewhere in the docs so it's not lost here?

**Reply by zzzeek** [reply]
but it's not a "F" aq !  the current place for "folks are curious about internals" is on the site: https://www.sqlalchemy.org/library.html#architecture  we can add more links there

**Reply by CaselIT** [reply]
I agree it's not a faq, unless we have a very low threshold for "frequently" :)
I think a faq like document in the docs is more appropriate for this. Something like an question/anwer page but for internal stuff of sqlalchemy. we could then link that from the architecture page.

**Reply by ShmuelTreiger** [reply]
Super cool! Thank you for sharing! And that detailed writeup! That's some nifty work you've been doing!

**Reply by CaselIT** [reply]
I've opted to add a wiki entry since it seems that's what fits the most: https://github.com/sqlalchemy/sqlalchemy/wiki/How-does-SQLAlchemy-keep-track-of-the-right-columns%3F

**Reply by zzzeek** [reply]
should we add an index entry to the wiki for things like this?

**Reply by CaselIT** [reply]
It may make sense

**Reply by CaselIT** [reply]
Do you also have other candidates for this new section of the wiki index?

**Reply by zzzeek** [reply]
i don't!  which is why im not too optimistic this document is going to be all that useful out by itself.

**Reply by CaselIT** [reply]
I may take a look at the wiki and organize it a bit better when I can spare the time
